### PR TITLE
executor, expression: rewrite builtin func left and right

### PR DESCRIPTION
**Pull request by SteveZhangBit**
Because most parts of the left and right function are the same, so I put them together.

### CODE REVIEW

File: expression/builtin_string.go [hunk 1]
```diff
@@ -295,37 +296,54 @@ type leftFunctionClass struct {
...
+
+	d, isNull, err = b.args[0].EvalString(row, b.ctx.GetSessionVars().StmtCtx)
+	if isNull || err != nil {
+		return d, isNull, errors.Trace(err)
 	}
-	str, err := args[0].ToString()
-	if err != nil {
-		return d, errors.Trace(err)
+	left, isNull, err = b.args[1].EvalInt(row, b.ctx.GetSessionVars().StmtCtx)
+	if terror.ErrorEqual(err, types.ErrTruncated) {
```
Comment (hanfei1991) [inline expression/builtin_string.go]: This err should be handled in cast funtion.

Comment (SteveZhangBit) [inline expression/builtin_string.go]: What do you mean by 'handle in cast function'? Change the code in 'builtinCastStringAsIntSig'?
What if we need to return the Truncated error?

File: expression/builtin_string.go [hunk 2]
```diff
@@ -295,37 +296,54 @@ type leftFunctionClass struct {
...
+		return nil, errors.Trace(err)
+	}
+	sig := &builtinLeftSig{baseStringBuiltinFunc{bf}}
 	return sig.setSelf(sig), errors.Trace(c.verifyArgs(args))
 }
 
+func (c *leftFunctionClass) inferRetType(args []Expression) *types.FieldType {
+	argTp := args[0].GetType()
+	tp := types.MergeFieldType(mysql.TypeVarString, argTp.Tp)
+	retType := types.NewFieldType(tp)
```
Comment (shenli) [inline expression/builtin_string.go]: Please set the Flen field of retType.

Comment (SteveZhangBit) [inline expression/builtin_string.go]: But the Flen field will change with the return value, like left("abc", 0) will get 0 and left("abc", 2) will get 2. How to handle such situation?

Comment (shenli) [inline expression/builtin_string.go]: Please refer to MySQL's rule:
https://github.com/mysql/mysql-server/blob/5.7/sql/item_strfunc.cc#L1688
https://github.com/mysql/mysql-server/blob/5.7/sql/item_strfunc.cc#L1668

Comment (SteveZhangBit) [inline expression/builtin_string.go]: The Flen value is related to the return value, so I set the value in the eval stage.

**Comment by XuHuaiyu** [comment]
@SteveZhangBit 
Please fix ci
can #3541 be closed?

**Comment by SteveZhangBit** [comment]
@XuHuaiyu 
I notice the error in ci, it's a golint error,  I will fix it soon.
I think #3541 could be closed :)

**Comment by SteveZhangBit** [comment]
@XuHuaiyu 
error fixed

File: expression/builtin_string.go [hunk 3]
```diff
@@ -293,89 +292,104 @@ type leftFunctionClass struct {
 }
 
 func (c *leftFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	sig := &builtinLeftSig{newBaseBuiltinFunc(args, ctx)}
+	tp := types.NewFieldType(mysql.TypeVarString)
+	tp.Charset, tp.Collate = charset.CharsetUTF8, charset.CollationUTF8
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: we can use args[0].getType().Flen as tp.Flen.

File: executor/executor_test.go [hunk 4]
```diff
@@ -810,6 +810,28 @@ func (s *testSuite) TestStringBuiltin(c *C) {
...
+	tk.MustExec(`insert into t values('2', 2, 2.3, "2017-01-01 12:01:01", "12:01:01")`)
+	result = tk.MustQuery("select left(a, 2), left(b, 2), left(c, 2), left(d, 2), left(e, 2) from t")
+	result.Check(testkit.Rows("2 2 2. 20 12"))
+	result = tk.MustQuery(`select left("abc", 0), left("abc", -1), left(NULL, 1), left("abc", NULL)`)
+	result.Check(testkit.Rows("  <nil> <nil>"))
+	result = tk.MustQuery(`select left("abc", "a"), left("abc", 1.9), left("abc", 1.2)`)
+	result.Check(testkit.Rows(" ab a"))
+
+	// for right
+	tk.MustExec("drop table if exists t")
```
Comment (XuHuaiyu) [inline executor/executor_test.go]: we can reuse the table created for `left`

File: expression/builtin_string.go [hunk 5]
```diff
@@ -277,89 +276,80 @@ type leftFunctionClass struct {
 }
 
 func (c *leftFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	sig := &builtinLeftSig{newBaseBuiltinFunc(args, ctx)}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpString, tpString, tpInt)
+	bf.tp.Flen = args[0].GetType().Flen
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
```
Comment (zz-jason) [inline expression/builtin_string.go]: `err` checking should be just after the function `newBaseBuiltinFuncWithTp` called

File: expression/builtin_string.go [hunk 6]
```diff
@@ -277,89 +276,80 @@ type leftFunctionClass struct {
...
 	baseFunctionClass
 }
 
 func (c *rightFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	sig := &builtinRightSig{newBaseBuiltinFunc(args, ctx)}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpString, tpString, tpInt)
+	bf.tp.Flen = args[0].GetType().Flen
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
```
Comment (zz-jason) [inline expression/builtin_string.go]: ditto

**Comment by SteveZhangBit** [comment]
@zz-jason 
Fixed

**Comment by hanfei1991** [comment]
@XuHuaiyu PTAL

File: expression/builtin_string.go [hunk 7]
```diff
@@ -277,89 +276,80 @@ type leftFunctionClass struct {
...
-	str, err := arg0.ToString()
-	if err != nil {
-		return d, errors.Trace(err)
-	}
-	sc := new(variable.StatementContext)
-	sc.IgnoreTruncate = true
-	length, err := arg1.ToInt64(sc)
-	if err != nil {
-		return d, errors.Trace(err)
+	d, isNull, err = b.args[0].EvalString(row, b.ctx.GetSessionVars().StmtCtx)
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: define sc := b.ctx.GetSessionVars().StmtCtx

thus reuse it in line 341

File: expression/builtin_string.go [hunk 8]
```diff
@@ -277,89 +276,80 @@ type leftFunctionClass struct {
...
-	if err != nil {
-		return types.Datum{}, errors.Trace(err)
-	}
-	str, err := args[0].ToString()
-	if err != nil {
-		return d, errors.Trace(err)
+func (b *builtinLeftSig) evalString(row []types.Datum) (d string, isNull bool, err error) {
+	var left int64
+
+	d, isNull, err = b.args[0].EvalString(row, b.ctx.GetSessionVars().StmtCtx)
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: ditto

File: expression/builtin_string_test.go [hunk 9]
```diff
@@ -216,40 +216,139 @@ func (s *testEvaluatorSuite) TestConcatWS(c *C) {
...
+	for _, t := range typecases {
+		f, err := fc.getFunction(t.args, s.ctx)
+		c.Assert(err, IsNil)
+		retType := f.getRetTp()
+		c.Assert(retType.Tp, Equals, t.retType.Tp)
+		c.Assert(retType.Charset, Equals, t.retType.Charset)
+		c.Assert(retType.Collate, Equals, t.retType.Collate)
+		c.Assert(retType.Flag, Equals, t.retType.Flag)
+	}
+}
```
Comment (XuHuaiyu) [inline expression/builtin_string_test.go]: Please add test for isDetermistic for left and right.

**Comment by SteveZhangBit** [comment]
@XuHuaiyu 
fixed

File: expression/builtin_string_test.go [hunk 10]
```diff
@@ -349,6 +353,10 @@ func (s *testEvaluatorSuite) TestRight(c *C) {
 		c.Assert(retType.Collate, Equals, t.retType.Collate)
```
Comment (XuHuaiyu) [inline expression/builtin_string_test.go]: @SteveZhangBit 
we can move these test cases which are used for testing type inferring
to plan/typeinfer_test.go which is more readable.

same as `space`

**Comment by hanfei1991** [comment]
@SteveZhangBit Please resolve conficts.

**Comment by SteveZhangBit** [comment]
@hanfei1991 
conflicts resolved

**Pull request merged**